File: app/src/main/java/io/datalive/myreferrer/ReferrerReceiver.java
package io.datalive.myreferrer;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

public class ReferrerReceiver extends BroadcastReceiver {
    public static final String ACTION_UPDATE_DATA = "ACTION_UPDATE_DATA";
    private static final String ACTION_INSTALL_REFERRER = "com.android.vending.INSTALL_REFERRER";
    //    private static final String KEY_2 = "k2";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent == null) {
            Log.e("ReferrerReceiver", "Intent is null");
            return;
        }
        if (!ACTION_INSTALL_REFERRER.equals(intent.getAction())) {
            Log.e("ReferrerReceiver", "Wrong action! Expected: " + ACTION_INSTALL_REFERRER + " but was: " + intent.getAction());
            return;
        }
        Bundle extras = intent.getExtras();
        if (intent.getExtras() == null) {
            Log.e("ReferrerReceiver", "No data in intent");
            return;
        }

        String referrer = extras.getString("referrer");

        if (referrer != null && !referrer.equals("")) {
            Intent main = new Intent(context, FirstInstallActivity.class).putExtra("referrer", referrer);
            main.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(main);
        }
    }
}

File: app/src/main/java/io/datalive/myreferrer/ReferrerReceiver.kt
package io.datalive.myreferrer

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent

class ReferrerReceiver : BroadcastReceiver() {

    companion object {
        private val ACTION_INSTALL_REFERRER = "com.android.vending.INSTALL_REFERRER"
    }

    override fun onReceive(context: Context, intent: Intent?) {
        intent?.let {
            when (intent.action) {
                ACTION_INSTALL_REFERRER -> {
                    intent.extras?.getString("referrer").let {
                        val intent = Intent(context, FirstInstallActivity::class.java)
                                .putExtra("referrer", it)
                                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                        context.startActivity(intent)
                    }
                }
            }
        }
    }
}
